src/Checkpointer.jl

```diff
@@ -498,4 +512,155 @@ function restore!(
+
+# Returns
+An iterator that yields objects that should be saved.
+
+# Notes
+- Objects returned by the iterator may reference the same memory.
+- Fields listed in `ignore` are not traversed.
+"""
+function CacheIterator(cache, ignore::Set{Symbol})
+    stack = []
+    push!(stack, cache)
+    return CacheIterator(stack, ignore)
+end
+
+"""
+    iterate(itr::CacheIterator)
+
+Advance the iterator to obtain the next field in the field hierarchy of the
+object. If no fields remain, `nothing` is returned. Otherwise, a 2-tuple of the
+next element and the new iteration state is returned.
+"""
+function Base.iterate(itr::CacheIterator)
+    stack = itr.stack
+    ignore = itr.ignore
+    while !isempty(stack)
+        obj = pop!(stack)
+        stop(obj) && continue
+        is_leaf(obj) && return (obj, nothing)
+        T = typeof(obj)
+        fields = filter(x -> !(x in ignore), fieldnames(T))
+        for p in fields
+            push!(stack, getfield(obj, p))
+        end
+    end
+    return nothing
+end
+
+# Functions needed to implement the iterator interface
+# see https://docs.julialang.org/en/v1/manual/interfaces/#man-interface-iteration
+Base.iterate(iter::CacheIterator, _) = iterate(iter)
+Base.IteratorEltype(::CacheIterator) = Base.EltypeUnknown()
+Base.IteratorSize(::CacheIterator) = Base.SizeUnknown()
+Base.isdone(itr::CacheIterator, state...) = isempty(itr.stack)
+
+"""
+    stop(::Type)
+
+Returns `true` to stop recursing at the given object and `false` to continue
+recursing through the fields of the object.
+
+This function is used by `CacheIterator`. This is useful to stop recursing and
+not return the object.
+"""
+function stop(
+    ::Union{
+        AbstractTimeVaryingInput,
+        ClimaComms.AbstractCommsContext,
+        ClimaComms.AbstractDevice,
+        UnionAll,
+        DataType,
+    },
+)
+    return true # Stop recursing here
+end
+
+# Catch all for everything else
+stop(obj) = false
+
+"""
+    is_leaf(::Type)
+
+Returns `true` if the object should be saved and `false` otherwise.
+
+This function is used by `CacheIterator`.
+"""
+is_leaf(::Union{CC.DataLayouts.AbstractData, AbstractArray}) = true # Needed for saving data
+
+# Needed for error handling
+is_leaf(::Union{StaticArrays.StaticArray, Number, UnitRange, LinRange, Symbol}) = true
+
+# Needed for error handling
+is_leaf(::Dict) = true
+
+# Needed for error handling
+is_leaf(::Union{Dates.DateTime, Dates.UTInstant, Dates.Millisecond}) = true
+
+# Catch all for everything else
+is_leaf(::Any) = false
+
+"""
+    CacheIterator(sim::Interfacer.ComponentModelSimulation)
+
+Create an iterator over the cache in `sim` of objects that should be
+saved.
+
+To use this constructor, the function `get_cache_ignore` must be implemented
+for `sim` which takes in `sim` and returns a set of symbols to ignore while
+iterating through the fields of the cache.
+"""
+function CacheIterator(sim::Interfacer.ComponentModelSimulation)
```

Comment:
I'm not sure, but maybe it's better to define this only for `AtmosModelSimulation` for now. I'm on the fence because it can technically be used for other components if `get_cache_ignore` is implemented, but that isn't the case right now and maybe dispatching on `ComponentModelSimulation` here will be misleading.

I'll leave it up to you if you want to restrict this to only atmos, or add a method of `get_cache_ignore` for `ComponentModelSimulation`, which errors and explains that the other checkpointing system should be used or the function needs to be extended.